onnxruntime/core/optimizer/qdq_transformer/selectors_actions/shared/utils.cc

```diff
@@ -55,6 +55,9 @@ static const OpVersionsAndSelector::OpVersionsMap GetConvOpVersionsMap() {
 static const OpVersionsAndSelector::OpVersionsMap GetMatMulOpVersionsMap() {
   return {{"MatMul", {}}};
 }
+static const OpVersionsAndSelector::OpVersionsMap GetGemmOpVersionsMap() {
```

Comment:
How can we share this QDQ selector across different EPs? for example I used node unit in our EP and I want to add GlobalAveragePool as a node unit also for our EPs? What we need to do here?

Reply:
thanks for the question.

The shared QDQ selectors here are mainly used for detecting a QDQ node unit group with associated targeted node.

For example, right now GlobalAveragePool operator is not registered in these shared OpVersionMap here. first step might be to register the operator here in this utils.cc file the similar way as for the other operators. (and for GlobalAveragePool this specific op, it can use a UnaryNodeGroupSelector so should be fine, no additional nodegroupselector needs to be implemented).

After this step, supposedly a QDQ GlobalAveragePool node unit is able to be selected.

The way to get a series of qdq selection results can see the usage here in this UT Test `QDQTransformerTests.QDQ_Selector_Test`:https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/optimizer/qdq_transformer_test.cc#L2145

The rest of the work I believe depends on the specific op support in your EP. And for integration from a single node op support -> qdq node unit one, can probably refer to nnapi qdq node unit integration.

Reply:
This should be the central place where all known QDQ groups are registered, so feel free to add GlobalAveragePool here. Guessing you can just add it to the list in GetUnaryOpVersionsMap as the input/output is the same as AveragePool aren't they? 

Within the EP you should use the utilities to get all the QDQ node groups in the model so they can be handled. You have to take all or none of the QDQ node group. 

[https://github.com/microsoft/onnxruntime/blob/bab9b80f1f2330d3a115e0abbb4d8278c2be3f44/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L113-L117](https://github.com/microsoft/onnxruntime/blob/bab9b80f1f2330d3a115e0abbb4d8278c2be3f44/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L113-L117)

Provide a helper when using partitioning utils like this: 

[https://github.com/microsoft/onnxruntime/blob/bab9b80f1f2330d3a115e0abbb4d8278c2be3f44/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L129](https://github.com/microsoft/onnxruntime/blob/bab9b80f1f2330d3a115e0abbb4d8278c2be3f44/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L129)

Up to the EP author to decide how best to implement the checks for whether something is supported. Do this check on the target node. For NNAPI it's via this: 

[https://github.com/microsoft/onnxruntime/blob/bab9b80f1f2330d3a115e0abbb4d8278c2be3f44/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L140-L142](https://github.com/microsoft/onnxruntime/blob/bab9b80f1f2330d3a115e0abbb4d8278c2be3f44/onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc#L140-L142). 

The CPU EP does it differently as it uses a SelectorActionTransformer. Most likely you'll do something similar to NNAPI.

Reply:
Yes, I did that to add GlobalAveragePool into UnaryNodeGroupSelector. and It worked. Then I encountered another op Gemm. I was thinking that should we add all Ops that should support here first, then at another level in the Helper code to add more ops support later?